File: ui/utils/__tests__/credentialSecret.test.ts
import { describe, expect, it } from 'vitest';
import { resolveCredentialAuthSecret, resolveCredentialPayload } from '../credentialSecret';

// The shapes below are the ones read-only production inspection found on
// layer5io/meshery-cloud issue #5918: ~36.9k Kubernetes credentials shaped
// {auth, cluster}, ~154 shaped {secret: '<string>'}, and 8 double-nested
// {credentialName, secret: {...}} rows written by Meshery UI's credential form.
// The canonical shape is what meshery/schemas declares and what Cloud is moving
// to. All four have to keep resolving.

describe('resolveCredentialPayload', () => {
  it('returns the canonical secret object as the payload', () => {
    const secret = { prometheusURL: 'https://prom.example' };
    expect(resolveCredentialPayload(secret)).toEqual(secret);
  });

  it('returns the kubernetes secret object as the payload', () => {
    const secret = {
      auth: { clusterToken: 'tok' },
      cluster: { server: 'https://k8s.example' },
    };
    expect(resolveCredentialPayload(secret)).toEqual(secret);
  });

  it('unwraps the legacy double-nested payload', () => {
    expect(
      resolveCredentialPayload({
        credentialName: 'kube-cred',
        secret: { auth: { clusterToken: 'tok' } },
      }),
    ).toEqual({ auth: { clusterToken: 'tok' } });
  });

  it('unwraps a legacy double-nested payload keyed by name', () => {
    expect(
      resolveCredentialPayload({ name: 'kube-cred', secret: { clusterName: 'cluster-a' } }),
    ).toEqual({ clusterName: 'cluster-a' });
  });

  it('unwraps the legacy string-valued secret', () => {
    expect(resolveCredentialPayload({ secret: 'super-secret' })).toBe('super-secret');
  });

  it('leaves a canonical payload that carries its own secret field alone', () => {
    const secret = { grafanaURL: 'https://grafana.example', secret: { nested: 'value' } };
    expect(resolveCredentialPayload(secret)).toEqual(secret);
  });

  it('ignores an inherited secret property when classifying the wrapper', () => {
    // Every own key is a wrapper key, so the own-key check alone cannot decide
    // this: only the `secret` probe can. With `'secret' in obj` the object would
    // be treated as a wrapper and unwrapped to the *prototype's* value, even
    // though it owns no `secret` at all.
    const payload = Object.create({ secret: { grafanaAPIKey: 'inherited' } });
    payload.name = 'grafana-cred';

    expect(resolveCredentialPayload(payload)).toBe(payload);
    expect(resolveCredentialAuthSecret(payload)).toBeUndefined();
  });

  it('passes through values that are not objects', () => {
    expect(resolveCredentialPayload(undefined)).toBeUndefined();
    expect(resolveCredentialPayload(null)).toBeNull();
    expect(resolveCredentialPayload('raw')).toBe('raw');
  });
});

describe('resolveCredentialAuthSecret', () => {
  it('reads the legacy string-valued secret', () => {
    expect(resolveCredentialAuthSecret({ secret: 'super-secret' })).toBe('super-secret');
    expect(
      resolveCredentialAuthSecret({ credentialName: 'prom-token', secret: 'super-secret' }),
    ).toBe('super-secret');
  });

  it('reads the canonical grafana API key', () => {
    expect(
      resolveCredentialAuthSecret({
        grafanaURL: 'https://grafana.example',
        grafanaAPIKey: 'canonical-key',
      }),
    ).toBe('canonical-key');
  });

  it('reads the grafana API key out of the legacy double-nested shape', () => {
    expect(
      resolveCredentialAuthSecret({
        credentialName: 'grafana-cred',
        secret: { grafanaURL: 'https://grafana.example', grafanaAPIKey: 'nested-key' },
      }),
    ).toBe('nested-key');
  });

  it('prefers the canonical field over a sibling legacy secret string', () => {
    expect(
      resolveCredentialAuthSecret({ grafanaAPIKey: 'canonical-key', secret: 'legacy-key' }),
    ).toBe('canonical-key');
  });

  it('reports no auth material for an anonymous canonical credential', () => {
    expect(resolveCredentialAuthSecret({ prometheusURL: 'https://prom.example' })).toBeUndefined();
    expect(
      resolveCredentialAuthSecret({
        credentialName: 'prom-cred',
        secret: { prometheusURL: 'https://prom.example' },
      }),
    ).toBeUndefined();
  });

  it('reports no auth material for a kubernetes credential', () => {
    expect(
      resolveCredentialAuthSecret({
        auth: { clusterToken: 'tok' },
        cluster: { server: 'https://k8s.example' },
      }),
    ).toBeUndefined();
  });

  it('reports no auth material for absent or non-object secrets', () => {
    expect(resolveCredentialAuthSecret(undefined)).toBeUndefined();
    expect(resolveCredentialAuthSecret(null)).toBeUndefined();
    expect(resolveCredentialAuthSecret({})).toBeUndefined();
  });
});

File: server/models/credential_secret_test.go
package models

import (
	"reflect"
	"testing"
)

// The shapes below are the ones read-only production inspection found on
// layer5io/meshery-cloud issue #5918: ~36.9k Kubernetes credentials shaped
// {auth, cluster}, ~154 shaped {secret: "<string>"}, and 8 double-nested
// {credentialName, secret: {...}} rows written by Meshery UI's credential form.
// The canonical shape is what meshery/schemas declares and what Cloud is moving
// to. All four have to keep resolving.

func TestCredentialPayload(t *testing.T) {
	tests := []struct {
		name   string
		secret map[string]interface{}
		want   map[string]interface{}
	}{
		{
			name:   "canonical prometheus payload is the secret map itself",
			secret: map[string]interface{}{"prometheusURL": "https://prom.example"},
			want:   map[string]interface{}{"prometheusURL": "https://prom.example"},
		},
		{
			name: "canonical kubernetes payload is the secret map itself",
			secret: map[string]interface{}{
				"clusterName":      "cluster-a",
				"clusterServerURL": "https://k8s.example",
				"auth":             map[string]interface{}{"clusterToken": "tok"},
			},
			want: map[string]interface{}{
				"clusterName":      "cluster-a",
				"clusterServerURL": "https://k8s.example",
				"auth":             map[string]interface{}{"clusterToken": "tok"},
			},
		},
		{
			name: "stored kubernetes shape is the secret map itself",
			secret: map[string]interface{}{
				"auth":    map[string]interface{}{"clusterToken": "tok"},
				"cluster": map[string]interface{}{"server": "https://k8s.example"},
			},
			want: map[string]interface{}{
				"auth":    map[string]interface{}{"clusterToken": "tok"},
				"cluster": map[string]interface{}{"server": "https://k8s.example"},
			},
		},
		{
			name: "legacy double-nested payload is unwrapped",
			secret: map[string]interface{}{
				"credentialName": "kube-cred",
				"secret": map[string]interface{}{
					"auth":    map[string]interface{}{"clusterToken": "tok"},
					"cluster": map[string]interface{}{"server": "https://k8s.example"},
				},
			},
			want: map[string]interface{}{
				"auth":    map[string]interface{}{"clusterToken": "tok"},
				"cluster": map[string]interface{}{"server": "https://k8s.example"},
			},
		},
		{
			name: "legacy double-nested payload keyed by name is unwrapped",
			secret: map[string]interface{}{
				"name":   "kube-cred",
				"secret": map[string]interface{}{"auth": map[string]interface{}{"clusterToken": "tok"}},
			},
			want: map[string]interface{}{"auth": map[string]interface{}{"clusterToken": "tok"}},
		},
		{
			name:   "legacy string shape carries no object payload",
			secret: map[string]interface{}{"secret": "super-secret"},
			want:   nil,
		},
		{
			name: "a payload carrying its own secret field is not unwrapped",
			secret: map[string]interface{}{
				"grafanaURL": "https://grafana.example",
				"secret":     map[string]interface{}{"nested": "value"},
			},
			want: map[string]interface{}{
				"grafanaURL": "https://grafana.example",
				"secret":     map[string]interface{}{"nested": "value"},
			},
		},
		{
			name:   "nil secret",
			secret: nil,
			want:   nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := CredentialPayload(tt.secret)
			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf("CredentialPayload() = %#v, want %#v", got, tt.want)
			}
		})
	}
}

func TestCredentialAuthSecret(t *testing.T) {
	tests := []struct {
		name   string
		secret map[string]interface{}
		want   string
	}{
		{
			name:   "legacy string shape",
			secret: map[string]interface{}{"secret": "super-secret"},
			want:   "super-secret",
		},
		{
			name:   "legacy string shape alongside a credential name",
			secret: map[string]interface{}{"credentialName": "prom-token", "secret": "super-secret"},
			want:   "super-secret",
		},
		{
			name: "canonical grafana payload",
			secret: map[string]interface{}{
				"grafanaURL":    "https://grafana.example",
				"grafanaAPIKey": "canonical-key",
			},
			want: "canonical-key",
		},
		{
			name: "legacy double-nested grafana payload",
			secret: map[string]interface{}{
				"credentialName": "grafana-cred",
				"secret": map[string]interface{}{
					"grafanaURL":    "https://grafana.example",
					"grafanaAPIKey": "nested-key",
				},
			},
			want: "nested-key",
		},
		{
			name:   "canonical prometheus payload is anonymous",
			secret: map[string]interface{}{"prometheusURL": "https://prom.example"},
			want:   "",
		},
		{
			name: "kubernetes payload carries no string auth material",
			secret: map[string]interface{}{
				"auth":    map[string]interface{}{"clusterToken": "tok"},
				"cluster": map[string]interface{}{"server": "https://k8s.example"},
			},
			want: "",
		},
		{
			name: "canonical field wins over a sibling legacy secret string",
			secret: map[string]interface{}{
				"grafanaAPIKey": "canonical-key",
				"secret":        "legacy-key",
			},
			want: "canonical-key",
		},
		{
			name:   "nil secret",
			secret: nil,
			want:   "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := CredentialAuthSecret(tt.secret); got != tt.want {
				t.Fatalf("CredentialAuthSecret() = %q, want %q", got, tt.want)
			}
		})
	}
}
